### PR TITLE
Add E2E tests for bi-directional scrolling in Thread

### PR DESCRIPTION
**Pull request by testableapple**
### 🔗 Issue Links

Resolves https://github.com/GetStream/ios-issues-tracking/issues/388

### 🎯 Goal

Add E2E tests for bi-directional scrolling in thread

### 📝 Summary

- all tests that we have for bi-directional scrolling in MessageList are copy-pasted to ThreadList
- mock server had some bugs around sending messages to threads, this should be fixed now

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

gif

### CODE REVIEW

File: TestTools/StreamChatTestMockServer/Robots/ParticipantRobot.swift [hunk 1]
```diff
@@ -310,20 +346,25 @@ public class ParticipantRobot {
...
         startTypingInThread()
         stopTypingInThread()
-
+        
+        let quotedMessage = toLastMessage ? server.lastMessageInThread : server.firstMessageInThread
+        let quotedMessageId = quotedMessage?[messageKey.id.rawValue] as? String
         let parentId = threadParentId ?? (server.lastMessage?[messageKey.id.rawValue] as? String)
         server.websocketMessage(
-            channelId: server.currentChannelId,
+//            channelId: server.currentChannelId,
```
Comment (nuno-vieira) [inline TestTools/StreamChatTestMockServer/Robots/ParticipantRobot.swift:357]: Should we delete this? Why is this commented out?

**Review by nuno-vieira — COMMENTED** [review]
Overall looks great 👌 Just added a small comment. And also, I have some suggestions, but I guess we can do this in a separate PR, but let me know what you think:

Would be nice to have the following scenarios tested, since we had some issues with it previously:

**Scenario 1 - Root message should only be visible in the last page**
1- Open channel
2- Open Thread with 30 messages
3- Parent message should not be visible
4- Scroll to top (Loads one more page)
5- Parent message should be visible

**Scenario 2 - Parent message should be visible if messagesCount equal page size**
1- Open channel
2- Open Thread with **25** messages (25 is important, should be the same as page size)
3- Parent message should NOT be visible
4- Scroll to top (Loads one more page)
5- Parent message should be visible

**Scenario 3 - Should be able to jump to root message when not in the list**
1- Open channel
2- Open Thread with 30 messages
4- Scroll to top (Loads one more page)
5- Quote reply root message
6- Close Thread
7- Open Thread
8- Jump to root message

File: TestTools/StreamChatTestMockServer/Robots/ParticipantRobot.swift (hunk 1 above)
Comment (testableapple) [reply TestTools/StreamChatTestMockServer/Robots/ParticipantRobot.swift:357]: Oops 🙈, fixed.

**Comment by testableapple** [comment]
> Overall looks great 👌 Just added a small comment. And also, I have some suggestions, but I guess we can do this in a separate PR, but let me know what you think:
> 
> Would be nice to have the following scenarios tested, since we had some issues with it previously:
> 
> **Scenario 1 - Root message should only be visible in the last page** 1- Open channel 2- Open Thread with 30 messages 3- Parent message should not be visible 4- Scroll to top (Loads one more page) 5- Parent message should be visible
> 
> **Scenario 2 - Parent message should be visible if messagesCount equal page size** 1- Open channel 2- Open Thread with **25** messages (25 is important, should be the same as page size) 3- Parent message should NOT be visible 4- Scroll to top (Loads one more page) 5- Parent message should be visible
> 
> **Scenario 3 - Should be able to jump to root message when not in the list** 1- Open channel 2- Open Thread with 30 messages 4- Scroll to top (Loads one more page) 5- Quote reply root message 6- Close Thread 7- Open Thread 8- Jump to root message

@nuno-vieira, I added these scenarios and a couple more, as discussed in DM 🙌
Running [cron-checks](https://github.com/GetStream/stream-chat-swift/actions/runs/4767305065) at the moment to make sure all tests pass on all supported iOS versions.

**Review by nuno-vieira — CHANGES_REQUESTED** [review]
LGTM! ✅

**Review by nuno-vieira — APPROVED** [review]
Miss Clicked :D

**Pull request merged**